File: Manual/LightField/Add-in and Automation SDK/Samples/VisualBasic Add-Ins/Data Sample/DataSample.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Markup
Imports System.AddIn
Imports System.IO
Imports System.AddIn.Pipeline
Imports PrincetonInstruments.LightField.AddIns

''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
'''''''''' This sample generates data in memory and dumps it into data views for display. ''''''''''
''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
Public Class DataSample
    Inherits AddInBase

    Public toolBarState_ As Nullable(Of Boolean)

    Public ReadOnly Property UISupport As UISupport
        Get
            Return UISupport.ApplicationToolBar
        End Get
    End Property

    Public Sub Activate(ByVal app As ILightFieldApplication)
        'Capture interface
        LightFieldApplication = app
        toolBarState_ = Nothing 'Forces it to be a button. True/False would make it a checkbox.
    End Sub

    Public Sub Deactivate()
    End Sub

    Public Overrides Sub UIApplicationToolBarExecute()
        GenerateRamps()
    End Sub
    Public Overrides Property UIApplicationToolBarIsChecked As Nullable(Of Boolean)
        Get
            Return toolBarState_
        End Get
        Set(ByVal value As Nullable(Of Boolean)) 'The button in LightField was pushed
            toolBarState_ = value
        End Set
    End Property

    Public Overrides ReadOnly Property UIApplicationToolBarTitle As String
        Get
            Return "Data Sample"
        End Get
    End Property

    Public Overrides ReadOnly Property UIApplicationToolBarToolTip As String
        Get
            Return "Generate Linear Ramps frames of data and display them as an example of how to create, modify, and display data"
        End Get
    End Property

    Public Overrides ReadOnly Property UIApplicationToolBarIsEnabled As Boolean
        Get
            Return True
        End Get
    End Property

    'Make a linear ramp of data and display it         
    Private Sub GenerateRamps()
        'Make a frame 200x400 pixels
        Dim frame1(200 * 400) As UShort
        For pix As Integer = 0 To 200 * 400 - 1 Step 1
            frame1(pix) = pix Mod 400
        Next
        'Make a frame 300x500 pixels with a 1k bias
        Dim frame2(300 * 500) As UShort
        For pix As Integer = 0 To 300 * 500 - 1 Step 1
            frame2(pix) = pix Mod 500 + 1000
        Next
        'Get the addin file manager
        Dim datamgr As IDataManager = LightFieldApplication.DataManager
        If datamgr IsNot Nothing Then
            Dim roi As RegionOfInterest = New RegionOfInterest(0, 0, 400, 200, 1, 1)
            Dim roi2 As RegionOfInterest = New RegionOfInterest(0, 0, 500, 300, 1, 1)
            'Simple single region
            Dim imageData As IImageDataSet = datamgr.CreateImageDataSet(frame1, roi, ImageDataFormat.MonochromeUnsigned16)
            'Complex Set contains two regions
            Dim rois = New RegionOfInterest() {roi, roi2}
            Dim buffers As New List(Of Array)
            buffers.Add(frame1)
            buffers.Add(frame2)
            Dim imageDataSets As IImageDataSet = datamgr.CreateImageDataSet(buffers, rois, ImageDataFormat.MonochromeUnsigned16)
            'Display
            Dim display As IDisplay = LightFieldApplication.DisplayManager
            If display IsNot Nothing Then
                'Select data file compare mode with 2 horizontal windows 
                display.ShowDisplay(DisplayLocation.ExperimentWorkspace, DisplayLayout.TwoHorizontal)
                Dim view As IDisplayViewer = Nothing

                'Modify the underlying data a little bit before displaying it.
                ModifyDataExample(imageData, 0)
                'Put the simple data in window 0
                view = display.GetDisplay(DisplayLocation.ExperimentWorkspace, 0)
                view.Display("SimpleRamp", imageData)

                'Modify the underlying data a little bit before displaying it.
                ModifyDataExample(imageDataSets, 1)
                'Put the complex data in window 1
                view = display.GetDisplay(DisplayLocation.ExperimentWorkspace, 1)
                view.Display("ComplexRamp", imageDataSets)
            End If
        End If
    End Sub

    'Change some of the values in the underlying IImageDataSet
    Private Sub ModifyDataExample(ByVal imageData As IImageDataSet, ByVal roiIdx As Integer)
        'Demostrate how to access the data and change it                
        Dim rois As RegionOfInterest() = imageData.Regions
        'Get the width & the height
        Dim h As Integer = rois(roiIdx).Height
        Dim w As Integer = rois(roiIdx).Width
        'Get sub image arrays from the data set 
        Dim row As IImageData = imageData.GetRow(roiIdx, 0, h / 2)
        Dim col As IImageData = imageData.GetColumn(roiIdx, 0, w / 2)
        Dim pix As IImageData = imageData.GetPixel(roiIdx, 0, h / 2, w / 2)
        'Create new arrays to replace data
        Dim rowData(w) As UShort
        Dim colData(h) As UShort
        Dim pixData(1) As UShort
        'Build up new data with values
        For x As Integer = 0 To w - 1
            rowData(x) = 5000 + x
        Next
        For y As Integer = 0 To h - 1
            colData(y) = 10000 + y
        Next
        pixData(0) = 65535 'Bright pixel
        'Push back to original buffer
        row.SetData(rowData)
        col.SetData(colData)
        pix.SetData(pixData)
    End Sub

End Class

File: Manual/LightField/Add-in and Automation SDK/Samples/CSharp Add-Ins/Data Sample/DataSample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.AddIn;
using System.AddIn.Pipeline;

using PrincetonInstruments.LightField.AddIns;

namespace LightFieldAddInSamples
{
    ///////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////
    [AddIn("Data Sample (C#)", Version = "1.0.0", Publisher = "Teledyne Princeton Instruments", Description = "Shows how to create, modify, and display data")]
    [QualificationData("IsSample", "True")]
    public class AddinDataSample : AddInBase, ILightFieldAddIn
    {
        bool? toolBarState_;
        ///////////////////////////////////////////////////////////////////////
        public UISupport UISupport { get { return UISupport.ApplicationToolBar; } }
        ///////////////////////////////////////////////////////////////////////
        public void Activate(ILightFieldApplication app)
        {
            // Capture Interface
            LightFieldApplication = app;
            toolBarState_ = null; // Forces it to be a button, true or false
                                  // would make it a checkbox
        }
        ///////////////////////////////////////////////////////////////////////
        public void Deactivate() { }
        ///////////////////////////////////////////////////////////////////////
        public override string UIApplicationToolBarTitle { get { return "Data Sample"; } }
        ///////////////////////////////////////////////////////////////////////
        public override bool? UIApplicationToolBarIsChecked
        {
            get { return toolBarState_; }

            // The Button In LightField Was Pushed
            set { toolBarState_ = value; }
        }
        ///////////////////////////////////////////////////////////////////////        
        public override void UIApplicationToolBarExecute() 
        {
            GenerateRamps(); 
        }
        ///////////////////////////////////////////////////////////////////////        
        public override string UIApplicationToolBarToolTip
        {
            get { return "Generate Linear Ramps frames of data and display them as an example of how to create, modify, and display data"; }
        }
        ///////////////////////////////////////////////////////////////////////        
        public override bool UIApplicationToolBarIsEnabled{ get { return true; } }
        ///////////////////////////////////////////////////////////////////////        
        //  Make a linear ramp of data and display it.         
        ///////////////////////////////////////////////////////////////////////
        private void GenerateRamps()
        {
            // Make A Frame 200x400 pixels
            ushort[] frame1 = new ushort[200 * 400];
            for (int pix = 0; pix < 200 * 400; pix++)
                frame1[pix] = (ushort)(pix % 400);

            // Make A Frame 300x500 pixels with a 1k bias
            ushort[] frame2 = new ushort[300 * 500];
            for (int pix = 0; pix < 300 * 500; pix++)
                frame2[pix] = (ushort)((pix % 500) + 1000);

            // Get the addin file manager
            var datamgr = LightFieldApplication.DataManager;
            if (datamgr != null)
            {
                RegionOfInterest roi  = new RegionOfInterest(0, 0, 400, 200, 1, 1);
                RegionOfInterest roi2 = new RegionOfInterest(0, 0, 500, 300, 1, 1);

                // Simple Single Region
                IImageDataSet imageData = datamgr.CreateImageDataSet(frame1, roi, ImageDataFormat.MonochromeUnsigned16);

                RegionOfInterest[] rois = { roi, roi2 };
                
                List<Array> buffers = new List<Array>();
                buffers.Add(frame1);
                buffers.Add(frame2);

                // Complex Set Containing Two Regions
                IImageDataSet imageDataSets = datamgr.CreateImageDataSet(buffers, rois, ImageDataFormat.MonochromeUnsigned16);

                IDisplay display = LightFieldApplication.DisplayManager;
                if (display != null)
                {
                    // Select Data File Compare Mode & 3 Vertical Windows 
                    display.ShowDisplay(DisplayLocation.ExperimentWorkspace, DisplayLayout.TwoHorizontal);
                    IDisplayViewer view = null;

                    // Modify the underlying data a little bit before displaying it.
                    ModifyDataExample(imageData, 0);

                    // Put the simple data in window 0
                    view = display.GetDisplay(DisplayLocation.ExperimentWorkspace, 0);
                    view.Display("SimpleRamp", imageData);

                    // Modify the underlying data a little bit before displaying it.
                    ModifyDataExample(imageDataSets, 1);

                    // Put the complex data in window 1
                    view = display.GetDisplay(DisplayLocation.ExperimentWorkspace, 1);
                    view.Display("ComplexRamp", imageDataSets);                    
                }                
            }           
        }
        ///////////////////////////////////////////////////////////////////////
        //  Change some of the values in the underlying IImageDataSet
        ///////////////////////////////////////////////////////////////////////
        void ModifyDataExample(IImageDataSet imageData, int roiIdx)
        {
            // Demostrate how to access the data and change it                
            var rois = imageData.Regions;

            // Get the width & the height
            int h = rois[roiIdx].Height;
            int w = rois[roiIdx].Width;

            // Get sub image arrays from the data set 
            IImageData row = imageData.GetRow(roiIdx, 0, h / 2);
            IImageData col = imageData.GetColumn(roiIdx, 0, w / 2);
            IImageData pix = imageData.GetPixel(roiIdx, 0, h / 2, w / 2);

            // Create new arrays to replace data
            ushort[] rowData = new ushort[w];
            ushort[] colData = new ushort[h];
            ushort[] pixData = new ushort[1];

            // Build up new data with values
            for (int x = 0; x < w; x++)
                rowData[x] = (ushort)(5000 + x);
            for (int y = 0; y < h; y++)
                colData[y] = (ushort)(10000 + y);

            // bright pixel
            pixData[0] = 65535;

            // Push back to original buffer
            row.SetData(rowData);
            col.SetData(colData);
            pix.SetData(pixData);        
        }
    }
}
